File: libs/language/src/lib/provider/NodeContentHandler.ts
import { BlueNode, NodeDeserializer } from '../model';
import { NodeToMapListOrValue } from '../utils';
import { yamlBlueParse } from '../../utils/yamlBlue';
import { JsonBlueValue, JsonBlueObject } from '../../schema';
import { OBJECT_BLUE_ID } from '../utils/Properties';

export interface StorageContentProcessor {
  prepareStorageNode(
    node: BlueNode,
    preprocessor: (node: BlueNode) => BlueNode,
  ): { blueId: string; node: BlueNode };

  prepareStorageNodeList(
    nodes: BlueNode[],
    preprocessor: (node: BlueNode) => BlueNode,
  ): { blueId: string; nodes: BlueNode[] };
}

export class ParsedContent {
  constructor(
    public readonly blueId: string,
    public readonly content: JsonBlueValue,
    public readonly isMultipleDocuments: boolean,
  ) {}
}

export class NodeContentHandler {
  private static readonly THIS_REFERENCE_PATTERN = /^this(#\d+)?$/;

  public static parseAndCalculateBlueId(
    content: string,
    preprocessor: (node: BlueNode) => BlueNode,
    storageProcessor: StorageContentProcessor,
  ): ParsedContent {
    let jsonNode: JsonBlueValue;

    try {
      const parsed = yamlBlueParse(content);
      if (parsed === undefined) {
        throw new Error();
      }
      jsonNode = parsed;
    } catch {
      throw new Error('Failed to parse content as YAML or JSON');
    }

    let blueId: string;
    let resultContent: JsonBlueValue;
    const isMultipleDocuments = Array.isArray(jsonNode) && jsonNode.length > 1;

    if (isMultipleDocuments) {
      const nodes = (jsonNode as JsonBlueValue[]).map((item: JsonBlueValue) =>
        NodeDeserializer.deserialize(item),
      );
      const prepared = storageProcessor.prepareStorageNodeList(
        nodes,
        preprocessor,
      );
      blueId = prepared.blueId;
      resultContent = prepared.nodes.map((node) =>
        NodeToMapListOrValue.get(node),
      );
    } else {
      const node = NodeDeserializer.deserialize(jsonNode);
      const prepared = storageProcessor.prepareStorageNode(node, preprocessor);
      blueId = prepared.blueId;
      resultContent = NodeToMapListOrValue.get(prepared.node);
    }

    return new ParsedContent(blueId, resultContent, isMultipleDocuments);
  }

  public static parseAndCalculateBlueIdForNode(
    node: BlueNode,
    preprocessor: (node: BlueNode) => BlueNode,
    storageProcessor: StorageContentProcessor,
  ): ParsedContent {
    const prepared = storageProcessor.prepareStorageNode(node, preprocessor);
    const jsonNode = NodeToMapListOrValue.get(prepared.node);

    return new ParsedContent(prepared.blueId, jsonNode, false);
  }

  public static parseAndCalculateBlueIdForNodeList(
    nodes: BlueNode[],
    preprocessor: (node: BlueNode) => BlueNode,
    storageProcessor: StorageContentProcessor,
  ): ParsedContent {
    if (!nodes || nodes.length === 0) {
      throw new Error('List of nodes cannot be null or empty');
    }

    const prepared = storageProcessor.prepareStorageNodeList(
      nodes,
      preprocessor,
    );
    const jsonNodes = prepared.nodes.map((node) =>
      NodeToMapListOrValue.get(node),
    );
    const isMultipleDocuments = nodes.length > 1;

    return new ParsedContent(prepared.blueId, jsonNodes, isMultipleDocuments);
  }

  public static resolveThisReferences(
    content: JsonBlueValue,
    currentBlueId: string,
    isMultipleDocuments: boolean,
  ): JsonBlueValue {
    return this.resolveThisReferencesRecursive(
      content,
      currentBlueId,
      isMultipleDocuments,
    );
  }

  private static resolveThisReferencesRecursive(
    content: JsonBlueValue,
    currentBlueId: string,
    isMultipleDocuments: boolean,
  ): JsonBlueValue {
    if (content && typeof content === 'object' && !Array.isArray(content)) {
      // Handle objects
      const result: JsonBlueObject = {};
      for (const [key, value] of Object.entries(content)) {
        if (typeof value === 'string') {
          if (
            key === OBJECT_BLUE_ID &&
            this.THIS_REFERENCE_PATTERN.test(value)
          ) {
            result[key] = this.resolveThisReference(
              value,
              currentBlueId,
              isMultipleDocuments,
            );
          } else {
            result[key] = value;
          }
        } else if (value && typeof value === 'object') {
          result[key] = this.resolveThisReferencesRecursive(
            value,
            currentBlueId,
            isMultipleDocuments,
          );
        } else {
          result[key] = value;
        }
      }
      return result;
    } else if (Array.isArray(content)) {
      // Handle arrays
      return content.map((element) => {
        if (element && typeof element === 'object') {
          return this.resolveThisReferencesRecursive(
            element,
            currentBlueId,
            isMultipleDocuments,
          );
        }
        return element;
      });
    }
    return content;
  }

  private static resolveThisReference(
    textValue: string,
    currentBlueId: string,
    isMultipleDocuments: boolean,
  ): string {
    if (isMultipleDocuments) {
      if (!textValue.startsWith('this#')) {
        throw new Error(
          "For multiple documents, 'this' references must include an index (e.g., 'this#0')",
        );
      }
      return currentBlueId + textValue.substring(4);
    } else {
      if (textValue === 'this') {
        return currentBlueId;
      } else {
        throw new Error(
          "For a single document, only 'this' is allowed as a reference, not 'this#<id>'",
        );
      }
    }
  }
}

File: references/java-sdk/src/main/java/blue/language/provider/NodeContentHandler.java
package blue.language.provider;

import blue.language.Blue;
import blue.language.blueid.BlueIdCalculator;
import blue.language.model.Node;
import blue.language.snapshot.ResolvedSnapshot;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.TextNode;

import java.util.List;
import java.util.function.Function;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import static blue.language.utils.UncheckedObjectMapper.JSON_MAPPER;
import static blue.language.utils.UncheckedObjectMapper.YAML_MAPPER;

public class NodeContentHandler {

    private static final Pattern THIS_REFERENCE_PATTERN = Pattern.compile("^this(#\\d+)?$");

    public static class ParsedContent {
        public final String blueId;
        public final JsonNode content;
        public final boolean isMultipleDocuments;

        public ParsedContent(String blueId, JsonNode content, boolean isMultipleDocuments) {
            this.blueId = blueId;
            this.content = content;
            this.isMultipleDocuments = isMultipleDocuments;
        }
    }

    public static ParsedContent parseAndCalculateBlueId(String content, Function<Node, Node> preprocessor) {
        JsonNode jsonNode;
        try {
            jsonNode = YAML_MAPPER.readTree(content);
        } catch (Exception e) {
            try {
                jsonNode = JSON_MAPPER.readTree(content);
            } catch (Exception ex) {
                throw new RuntimeException("Failed to parse content as YAML or JSON", ex);
            }
        }

        String blueId;
        boolean isMultipleDocuments = jsonNode.isArray() && jsonNode.size() > 1;

        if (isMultipleDocuments) {
            List<Node> nodes = StreamSupport.stream(jsonNode.spliterator(), false)
                    .map(item -> JSON_MAPPER.convertValue(item, Node.class))
                    .map(preprocessor)
                    .collect(Collectors.toList());
            blueId = BlueIdCalculator.calculateSemanticBlueId(nodes);
            jsonNode = JSON_MAPPER.valueToTree(nodes);
        } else {
            Node node = JSON_MAPPER.convertValue(jsonNode, Node.class);
            node = preprocessor.apply(node);
            blueId = BlueIdCalculator.calculateSemanticBlueId(node);
            jsonNode = JSON_MAPPER.valueToTree(node);
        }

        return new ParsedContent(blueId, jsonNode, isMultipleDocuments);
    }

    public static ParsedContent parseAndCalculateBlueId(Node node, Function<Node, Node> preprocessor) {
        String blueId;
        Node preprocessedNode = preprocessor.apply(node);
        blueId = BlueIdCalculator.calculateSemanticBlueId(preprocessedNode);
        JsonNode jsonNode = JSON_MAPPER.valueToTree(preprocessedNode);

        return new ParsedContent(blueId, jsonNode, false);
    }

    public static ParsedContent parseAndCalculateBlueId(List<Node> nodes, Function<Node, Node> preprocessor) {
        if (nodes == null || nodes.isEmpty()) {
            throw new IllegalArgumentException("List of nodes cannot be null or empty");
        }

        List<Node> preprocessedNodes = nodes.stream()
                .map(preprocessor)
                .collect(Collectors.toList());

        String blueId = BlueIdCalculator.calculateSemanticBlueId(preprocessedNodes);
        JsonNode jsonNode = JSON_MAPPER.valueToTree(preprocessedNodes);
        boolean isMultipleDocuments = nodes.size() > 1;

        return new ParsedContent(blueId, jsonNode, isMultipleDocuments);
    }

    public static ParsedContent parseAndCalculateSemanticBlueId(String content, Blue blue) {
        JsonNode jsonNode;
        try {
            jsonNode = YAML_MAPPER.readTree(content);
        } catch (Exception e) {
            try {
                jsonNode = JSON_MAPPER.readTree(content);
            } catch (Exception ex) {
                throw new RuntimeException("Failed to parse content as YAML or JSON", ex);
            }
        }

        if (jsonNode.isArray() && jsonNode.size() > 1) {
            List<Node> nodes = StreamSupport.stream(jsonNode.spliterator(), false)
                    .map(item -> JSON_MAPPER.convertValue(item, Node.class))
                    .collect(Collectors.toList());
            return parseAndCalculateSemanticBlueId(nodes, blue);
        }

        Node node = JSON_MAPPER.convertValue(jsonNode, Node.class);
        return parseAndCalculateSemanticBlueId(node, blue);
    }

    public static ParsedContent parseAndCalculateSemanticBlueId(Node node, Blue blue) {
        ResolvedSnapshot snapshot = blue.resolveToSnapshot(node);
        JsonNode canonical = JSON_MAPPER.valueToTree(snapshot.canonicalRoot().toNode());
        return new ParsedContent(snapshot.rootBlueId(), canonical, false);
    }

    public static ParsedContent parseAndCalculateSemanticBlueId(List<Node> nodes, Blue blue) {
        if (nodes == null || nodes.isEmpty()) {
            throw new IllegalArgumentException("List of nodes cannot be null or empty");
        }

        List<ResolvedSnapshot> snapshots = nodes.stream()
                .map(blue::resolveToSnapshot)
                .collect(Collectors.toList());

        List<Node> canonicalNodes = snapshots.stream()
                .map(snapshot -> snapshot.canonicalRoot().toNode())
                .collect(Collectors.toList());

        String blueId = BlueIdCalculator.calculateSemanticBlueId(canonicalNodes);
        JsonNode canonicalList = JSON_MAPPER.valueToTree(canonicalNodes);
        return new ParsedContent(blueId, canonicalList, canonicalNodes.size() > 1);
    }

    public static JsonNode resolveThisReferences(JsonNode content, String currentBlueId, boolean isMultipleDocuments) {
        return resolveThisReferencesRecursive(content, currentBlueId, isMultipleDocuments);
    }

    private static JsonNode resolveThisReferencesRecursive(JsonNode content, String currentBlueId, boolean isMultipleDocuments) {
        if (content.isObject()) {
            ObjectNode objectNode = (ObjectNode) content;
            objectNode.fields().forEachRemaining(entry -> {
                JsonNode value = entry.getValue();
                if (value.isTextual()) {
                    String textValue = value.asText();
                    if (THIS_REFERENCE_PATTERN.matcher(textValue).matches()) {
                        String newValue = resolveThisReference(textValue, currentBlueId, isMultipleDocuments);
                        objectNode.set(entry.getKey(), new TextNode(newValue));
                    }
                } else if (value.isObject() || value.isArray()) {
                    objectNode.set(entry.getKey(), resolveThisReferencesRecursive(value, currentBlueId, isMultipleDocuments));
                }
            });
            return objectNode;
        } else if (content.isArray()) {
            ArrayNode arrayNode = (ArrayNode) content;
            for (int i = 0; i < arrayNode.size(); i++) {
                JsonNode element = arrayNode.get(i);
                if (element.isTextual()) {
                    String textValue = element.asText();
                    if (THIS_REFERENCE_PATTERN.matcher(textValue).matches()) {
                        String newValue = resolveThisReference(textValue, currentBlueId, isMultipleDocuments);
                        arrayNode.set(i, new TextNode(newValue));
                    }
                } else if (element.isObject() || element.isArray()) {
                    arrayNode.set(i, resolveThisReferencesRecursive(element, currentBlueId, isMultipleDocuments));
                }
            }
            return arrayNode;
        }
        return content;
    }

    private static String resolveThisReference(String textValue, String currentBlueId, boolean isMultipleDocuments) {
        if (isMultipleDocuments) {
            if (!textValue.startsWith("this#")) {
                throw new IllegalArgumentException("For multiple documents, 'this' references must include an index (e.g., 'this#0')");
            }
            return currentBlueId + textValue.substring(4);
        } else {
            if (textValue.equals("this")) {
                return currentBlueId;
            } else {
                throw new IllegalArgumentException("For a single document, only 'this' is allowed as a reference, not 'this#<id>'");
            }
        }
    }
}